libyul/optimiser/StructuralSimplifier.cpp

```diff
@@ -0,0 +1,121 @@
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include <libyul/optimiser/StructuralSimplifier.h>
+#include <libyul/optimiser/Semantics.h>
+#include <libyul/AsmData.h>
+#include <libdevcore/CommonData.h>
+
+using namespace std;
+using namespace dev;
+using namespace yul;
+
+void StructuralSimplifier::operator()(Block& _block)
+{
+	iterateReplacing(
+		_block.statements,
+		[&](Statement& _stmt) -> boost::optional<vector<Statement>>
+		{
+			if (_stmt.type() == typeid(If))
+			{
+				auto& ifStmt = boost::get<If>(_stmt);
+				if (ifStmt.body.statements.empty())
+				{
+					if (MovableChecker(*ifStmt.condition).movable())
+						return {{}};
+					else
+						return {{ExpressionStatement{ifStmt.location, FunctionalInstruction{
+							ifStmt.location,
+							solidity::Instruction::POP,
+							{std::move(*ifStmt.condition)}
+						}}}};
+				}
+				if (ifStmt.condition->type() == typeid(Literal))
+				{
+					auto const& condition = boost::get<Literal>(*ifStmt.condition);
+					if (isTrue(condition))
+					{
+						(*this)(ifStmt.body);
+						return {{std::move(ifStmt.body.statements)}};
+					}
+					else if (isFalse(condition))
+						return {{}};
+				}
+			}
+			else if (_stmt.type() == typeid(Switch))
+			{
+				auto& switchStmt = boost::get<Switch>(_stmt);
+				if (switchStmt.cases.size() == 1)
+				{
+					auto& switchCase = switchStmt.cases.front();
+					(*this)(switchCase.body);
+					if (switchCase.value)
+						return {{If{
+							std::move(switchStmt.location),
+							make_shared<Expression>(
+								FunctionalInstruction{
+									locationOf(*switchStmt.expression),
+									solidity::Instruction::EQ,
+									{*switchStmt.expression, std::move(*switchCase.value)}
+								}
+						), std::move(switchCase.body)}}};
+					else
+						return {{ExpressionStatement{
+										locationOf(*switchStmt.expression),
+										FunctionalInstruction{
+											locationOf(*switchStmt.expression),
+											solidity::Instruction::POP,
+											{*switchStmt.expression}
+										}
+						}, std::move(switchCase.body)}};
+				}
+			}
+			else if (_stmt.type() == typeid(ForLoop))
+			{
+				auto& forLoop = boost::get<ForLoop>(_stmt);
+				if (
+					forLoop.condition->type() == typeid(Literal) &&
+					isFalse(boost::get<Literal>(*forLoop.condition))
+				)
+				{
+					for (auto& stmt: forLoop.pre.statements)
+						visit(stmt);
+					return {{std::move(forLoop.pre.statements)}};
+				}
+			}
+			visit(_stmt);
+			return {};
+		}
+	);
+}
+
+bool StructuralSimplifier::isTrue(Literal const& _literal)
+{
+	static YulString const trueString("true");
+	return
+		(_literal.kind == LiteralKind::Boolean && _literal.value == trueString) ||
+		(_literal.kind == LiteralKind::Number && u256(_literal.value.str()) != u256(0))
```

Comment:
This can throw exceptions. Perhaps it makes sense to combine this code with the one in SimplificationRules:208 (`Pattern:d()`) and export it into a utility.